skfp/similarities/tanimoto.py

```diff
@@ -0,0 +1,350 @@
+
+    vec_b : {ndarray, sparse matrix}
+        Second binary input array or sparse matrix.
+
+    Returns
+    ----------
+    distance : float
+        Tanimoto distance between A and vec_b.
+
+    Examples
+    ----------
+    >>> from skfp.similarities import tanimoto_count_distance
+    >>> vec_a = np.array([7, 1, 1])
+    >>> vec_b = np.array([7, 1, 2])
+    >>> dist = tanimoto_count_distance(vec_a, vec_b)
+    >>> dist
+    0.02
+
+    >>> vec_a = csr_array(([7, 1, 1]))
+    >>> vec_b = csr_array(([7, 1, 2]))
+    >>> dist = tanimoto_count_distance(vec_a, vec_b)
+    >>> dist
+    0.02
+    """
+
+    return 1 - tanimoto_count_similarity(vec_a, vec_b)
+
+
+def _tanimoto_binary_numpy(vec_a: np.ndarray, vec_b: np.ndarray) -> float:
+    """
+    Calculates the Tanimoto similarity between two binary numpy arrays.
+    """
+    _check_nan(vec_a)
+    _check_nan(vec_b)
+
+    if vec_a.sum() == 0 == vec_b.sum():
+        return 1.0
+
+    dot_ab: float = np.dot(vec_a, vec_b)
+    dot_aa: float = np.sum(vec_a)
+    dot_bb: float = np.sum(vec_b)
+
+    denominator: float = dot_aa + dot_bb - dot_ab
+    _check_zero_denominator(denominator)
+
+    tanimoto_similarity = dot_ab / denominator
+    return tanimoto_similarity
+
+
+def _tanimoto_binary_scipy(vec_a: csr_array, vec_b: csr_array) -> float:
+    """
+    Calculates the Tanimoto similarity between two binary scipy arrays.
+    """
+    _check_nan(vec_a)
+    _check_nan(vec_b)
+
+    if np.sum(vec_a) == 0 and np.sum(vec_b) == 0:
+        return 1.0
+
+    a_and_b: float = vec_a.multiply(vec_b).sum()
+    a_sum: float = vec_a.sum()
+    b_sum: float = vec_b.sum()
+
+    denominator: float = a_sum + b_sum - a_and_b
+    _check_zero_denominator(denominator)
+
+    tanimoto_sim: float = a_and_b / denominator
+
+    return tanimoto_sim
+
+
+def _tanimoto_count_numpy(vec_a: np.ndarray, vec_b: np.ndarray) -> float:
+    """
+    Calculates the Tanimoto similarity between two count data numpy arrays.
+    """
+    _check_nan(vec_a)
+    _check_nan(vec_b)
+
+    if np.sum(vec_a) == 0 and np.sum(vec_b) == 0:
+        return 1.0
+
+    dot_ab: float = np.dot(vec_a, vec_b)
+    dot_aa: float = np.dot(vec_a, vec_a)
+    dot_bb: float = np.dot(vec_b, vec_b)
+
+    denominator: float = dot_aa + dot_bb - dot_ab
+    _check_zero_denominator(denominator)
+
+    tanimoto_sim: float = dot_ab / denominator
+
+    return tanimoto_sim
+
+
+def _tanimoto_count_scipy(vec_a: csr_array, vec_b: csr_array) -> float:
+    """
+    Calculates the Tanimoto similarity between two count data scipy arrays.
+    """
+    _check_nan(vec_a)
+    _check_nan(vec_b)
+
+    if np.sum(vec_a) == 0 and np.sum(vec_b) == 0:
+        return 1.0
+
+    dot_ab: float = vec_a.multiply(vec_b).sum()
+    dot_aa: float = vec_a.multiply(vec_a).sum()
+    dot_bb: float = vec_b.multiply(vec_b).sum()
+
+    denominator: float = dot_aa + dot_bb - dot_ab
+    _check_zero_denominator(denominator)
+
+    tanimoto_sim: float = dot_ab / denominator
```

Comment:
I think that if you use `__mul__` instead of `multiply`, you can use the same function for both NumPy and SciPy

Reply:
We need to keep these methods separate since we want to use JIT with the NumPy one